# Manipulate integral and floating point numbers in C#

*In this tutorial about numeric types, you'll use Jupyter notebooks to learn C# interactively. You're going to write C# code and see the results of compiling and running your code directly in the notebook.*

It contains a series of lessons that explore numbers and math operations in C#. These lessons teach you the fundamentals of the C# language.

## Working with integer math

Run the following cell:


In [1]:
int a = 18;
int b = 6;
int c = a + b;
Console.WriteLine(c);


The below script needs to be able to find the current output cell; this is an easy method to get it.

24


You've seen one of the fundamental math operations with integers. The `int` type represents an **integer**, a positive or negative whole number. You use the `+` symbol for addition. Other common mathematical operations for integers include:

- `-` for subtraction
- `*` for multiplication
- `/` for division

Start by exploring those different operations. Modify the third line to try each of these operations. After each edit, select the **Run** button.

- Subtraction: `int c = a - b;`
- Multiplication: `int c = a * b;`
- Division: `int c = a / b;`
```

You can also experiment by writing multiple mathematics operations in the same line, if you'd like.

> As you explore C# (or any programming language), you'll make mistakes when you write code. The **compiler** will find those errors and report them to you. When the output contains error messages, look closely at the example code, and the code in the interactive window to see what to fix. That exercise will help you learn the structure of C# code.

## Explore order of operations

The C# language defines the precedence of different mathematics operations with rules consistent with the rules you learned in mathematics. Multiplication and division take precedence over addition and subtraction. Explore that by running the following cell:


In [3]:
int a = 5;
int b = 4;
int c = 2;
int d = a + b * c;
Console.WriteLine(d);

13


The output demonstrates that the multiplication is performed before the addition.

You can force a different order of operation by adding parentheses around the operation or operations you want performed first. Modify the fourth line in the cell above to force the addition to be performed first: `int d = (a + b) * c;`

Explore more by combining many different operations. Replace the fourth line above with something like this. `int d = (a + b) - 6 * c + (12 * 4) / 3 + 12;`

You may have noticed an interesting behavior for integers. Integer division always produces an integer result, even when you'd expect the result to include a decimal or fractional portion.

## Explore integer precision and limits

If you haven't seen this behavior, try the following cell:

In [5]:
int a = 7;
int b = 4;
int c = 3;
int d = (a + b) / c;
int e = (a + b) % c;
Console.WriteLine($"quotient: {d}");
Console.WriteLine($"remainder: {e}");

quotient: 3
remainder: 2


That last sample showed you that integer division truncates the result. It showed how you can get the **remainder** by using the **remainder** operator, the `%` character.

The C# integer type differs from mathematical integers in one other way: the `int` type has minimum and maximum limits. Run this cell to see those limits:

In [7]:
int max = int.MaxValue;
int min = int.MinValue;
Console.WriteLine($"The range of integers is {min} to {max}");

The range of integers is -2147483648 to 2147483647
-2147483646


If a calculation produces a value that exceeds those limits, you have an **underflow** or **overflow** condition. The answer appears to wrap from one limit to the other. Add these two lines to the preceding cell to see an example:

```csharp
int what = max + 3;
Console.WriteLine($"An example of overflow: {what}");
```

Notice that the answer is very close to the minimum (negative) integer. It's the same as `min + 2`. The addition operation **overflowed** the allowed values for integers. The answer is a very large negative number because an overflow "wraps around" from the largest possible integer value to the smallest.

There are other numeric types with different limits and precision that you would use when the `int` type doesn't meet your needs. Let's explore those types of numbers next.

## Work with the double type

The `double` numeric type represents a double-precision floating point number. Those terms may be new to you. A **floating point** number is useful to represent non-integral numbers that may be very large or small in magnitude. **Double-precision** is a relative term that describes the numbers of binary digits used to store the value. **Double precision** number have twice the number of binary digits as **single-precision**. On modern computers, it is more common to use double precision than single precision numbers. **Single precision** numbers are declared using the `float` keyword. Let's explore. Try the following cell:

In [8]:
double a = 5;
double b = 4;
double c = 2;
double d = (a + b) / c;
Console.WriteLine(d);

4.5


Notice that the answer includes the decimal portion of the quotient. Try a slightly more complicated expression with doubles:

In [9]:
double a = 19;
double b = 23;
double c = 8;
double d = (a + b) / c;
Console.WriteLine(d);

5.25


The range of a double value is much greater than integer values. Try the following cell:

In [11]:
double max = double.MaxValue;
double min = double.MinValue;
Console.WriteLine($"The range of double is {min} to {max}");

The range of double is -1.7976931348623157E+308 to 1.7976931348623157E+308


These values are printed out in scientific notation. The number to the left of the `E` is the significand. The number to the right is the exponent,
as a power of 10.
    
Just like decimal numbers in math, doubles in C# can have rounding errors. Try this cell:

In [12]:
double third = 1.0 / 3.0;
Console.WriteLine(third);

0.3333333333333333


You know that `0.3` is `3/10` and not exactly the same as `1/3`. Similarly, `0.33` is `33/100`. That's closer to `1/3`, but still not exact. 

***Challenge***

Try other calculations with large numbers, small numbers, multiplication, and division using the `double` type.  Try more complicated calculations.